### PR TITLE
Add List<ColumnMetadata> parameter to SystemAccessControl.checkCanSelectFromColumns

### PR DESCRIPTION
**Pull request by iirekm**

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java [hunk 1]
```diff
@@ -264,7 +264,7 @@ public void checkCanRenameColumn(ConnectorSecurityContext context, SchemaTableNa
     }
 
     @Override
-    public void checkCanSelectFromColumns(ConnectorSecurityContext context, SchemaTableName tableName, Set<String> columnNames)
+    public void checkCanSelectFromColumnsWithMetadata(ConnectorSecurityContext context, SchemaTableName tableName, Set<ColumnMetadata> columns)
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java:267]: I think we should sacrifice backward compatibility and change the parameter type without changing method name (adding new method)

@findepi @electrum what do you think?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java [hunk 2]
```diff
@@ -264,7 +264,7 @@ public void checkCanRenameColumn(ConnectorSecurityContext context, SchemaTableNa
     }
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java:264]: Please describe the motivation behind this change

File: presto-spi/src/test/java/io/prestosql/spi/connector/ColumnTestingUtil.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package io.prestosql.spi.connector;
```
Comment (kokosing) [inline presto-spi/src/test/java/io/prestosql/spi/connector/ColumnTestingUtil.java:14]: copyright

File: presto-spi/src/test/java/io/prestosql/spi/connector/ColumnTestingUtil.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.spi.connector;
+
+import io.prestosql.spi.type.VarcharType;
+
+public class ColumnTestingUtil
```
Comment (kokosing) [inline presto-spi/src/test/java/io/prestosql/spi/connector/ColumnTestingUtil.java:18]: TestingColumnMetadata and please make it final

File: presto-main/src/main/java/io/prestosql/sql/analyzer/Analyzer.java [hunk 5]
```diff
@@ -84,14 +87,24 @@ public Analysis analyze(Statement statement, boolean isDescribe)
...
+                                    accessControlInfo.getSecurityContext(session.getRequiredTransactionId(), session.getQueryId()),
+                                    tableName,
+                                    columns.stream().flatMap(columnName -> getColumnMetadata(tableHandle, columnName).stream()).collect(toSet())));
+                }));
         return analysis;
     }
 
+    private Optional<ColumnMetadata> getColumnMetadata(TableHandle tableHandle, String columnName)
+    {
+        // TODO why sometimes .get(columnName) gives null? (eg in TestRowFilter.testView)
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/Analyzer.java:103]: I think we need to fix `TestRowFilter` to not to return null. Maybe it is using some custom connector or local query runner

File: presto-main/src/main/java/io/prestosql/sql/analyzer/Analyzer.java [hunk 6]
```diff
@@ -84,14 +87,24 @@ public Analysis analyze(Statement statement, boolean isDescribe)
 
         // check column access permissions for each table
         analysis.getTableColumnReferences().forEach((accessControlInfo, tableColumnReferences) ->
-                tableColumnReferences.forEach((tableName, columns) ->
-                        accessControlInfo.getAccessControl().checkCanSelectFromColumns(
-                                accessControlInfo.getSecurityContext(session.getRequiredTransactionId(), session.getQueryId()),
-                                tableName,
-                                columns)));
+                tableColumnReferences.forEach((tableName, columns) -> {
+                    // TODO why getTableHandle sometimes gives empty()? (eg in TestAnalyzer.testExistingRecursiveView)
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/Analyzer.java:91]: I think we need to fix `TestAnalyzer` instead.

Comment (iirekm) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/Analyzer.java:91]: fixing in Analyzer was the easiest what I could do (1 place); the only alternative was fixing StatementAnalyzer and Analysis (2 places)

File: presto-main/src/main/java/io/prestosql/sql/analyzer/Analyzer.java (hunk 5 above)
Comment (iirekm) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/Analyzer.java:103]: I have no idea how, somebody must help me

**Comment by martint** [comment]
What’s the motivation for this change?

This is tricky, because column metadata is only defined for tables. We don’t have any provisions in the engine or in connectors to provide column metadata for view columns. Incidentally, that’s why you’re seeing missing TableHandles in parts of your code. That happens when the table reference is a view.

**Comment by iirekm** [comment]
> What’s the motivation for this change?

we need this change for security: only some users can display columns matching some criteria (and we need ColumnMetadata to detect those criteria)

**Comment by kokosing** [comment]
> What’s the motivation for this change?

The hive tables have `$path` column which contains sensitive information and in general it is not part of the table ddl. We typically require ownership like privileges to execute `SHOW CREATE TABLE` where user can potentially see the table location. With this change we will be able to implement access control and require ownership to access hidden tables.

In general I am mostly concern about hidden tables and properties. 

Next step is to pass table and schema metadata to `CREATE` statements. So access control can check what properties are used and if user is allowed to set them that way.

> We don’t have any provisions in the engine or in connectors to provide column metadata for view columns

I think we could have artificial very simple column metadata for views with no properties and no hidden columns.

**Comment by electrum** [comment]
The title for the commit message is too long. Notice how GitHub wraps it. Please limit to 50 characters. We follow this guide: https://chris.beams.io/posts/git-commit/

**Comment by electrum** [comment]
Using `ColumnMetadata` seems like the wrong approach here. The check shouldn't be relying on the fact that the column is hidden. Rather, it should know specifically about the `$path` column for the Hive connector.

**Comment by kokosing** [comment]
@iirekm I think we can close it.

**Pull request closed without merging**